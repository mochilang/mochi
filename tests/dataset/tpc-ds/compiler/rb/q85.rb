# Generated by Mochi compiler v0.10.27 on 2025-07-17T06:41:41Z
require 'ostruct'

def _avg(v)
  list = nil
  if defined?(MGroup) && v.is_a?(MGroup)
    list = v.Items
  elsif v.is_a?(Array)
    list = v
  elsif v.respond_to?(:to_a)
    list = v.to_a
  end
  return 0 if !list || list.empty?
  list.sum(0.0) / list.length
end
def _json(v)
  require 'json'
  obj = v
  if v.is_a?(Array)
    obj = v.map { |it| it.respond_to?(:to_h) ? it.to_h : it }
  elsif v.respond_to?(:to_h)
    obj = v.to_h
  end
  puts(JSON.generate(obj))
end

$web_returns = [OpenStruct.new(qty: 60, cash: 20.0, fee: 1.0), OpenStruct.new(qty: 100, cash: 30.0, fee: 2.0), OpenStruct.new(qty: 95, cash: 25.0, fee: 3.0)]
$result = _avg((($web_returns)).map { |r| r.qty })
_json($result)
raise "expect failed" unless ($result == 85.0)
