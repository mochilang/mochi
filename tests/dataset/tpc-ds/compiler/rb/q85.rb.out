require "ostruct"

def _json(v)
  require "json"
  obj = v
  if v.is_a?(Array)
    obj = v.map { |it| it.respond_to?(:to_h) ? it.to_h : it }
  elsif v.respond_to?(:to_h)
    obj = v.to_h
  end
  puts(JSON.generate(obj))
end

web_returns = [OpenStruct.new(qty: 60, cash: 20.0, fee: 1.0), OpenStruct.new(qty: 100, cash: 30.0, fee: 2.0), OpenStruct.new(qty: 95, cash: 25.0, fee: 3.0)]
result = (((web_returns.map { |r| r.qty }).length > 0) ? (web_returns.map { |r| r.qty }).sum(0.0) / (web_returns.map { |r| r.qty }).length : 0)
_json(result)
raise "expect failed" unless result == 85.0
